astroquery/nrao/core.py

```diff
@@ -195,6 +199,87 @@ def _args_to_payload(self, **kwargs):
+
+        Parameters
+        ----------
+        username : str, optional
+            Username to the NRAO archive. If not given, it should be specified
+            in the config file.
+        store_password : bool, optional
+            Stores the password securely in your keyring. Default is False.
+        reenter_password : bool, optional
+            Asks for the password even if it is already stored in the
+            keyring. This is the way to overwrite an already stored passwork
+            on the keyring. Default is False.
+        """
+
+        # Developer notes:
+        # Login via https://my.nrao.edu/cas/login
+        # # this can be added to auto-redirect back to the query tool: ?service=https://archive.nrao.edu/archive/advquery.jsp
+
+        if username is None:
+            if self.USERNAME == "":
+                raise LoginError("If you do not pass a username to login(), "
+                                 "you should configure a default one!")
+            else:
+                username = self.USERNAME
+
+        # Check if already logged in
+        loginpage = self._request("GET", "https://my.nrao.edu/cas/login",
+                                  cache=False)
+        root = BeautifulSoup(loginpage.content, 'html5lib')
+        if root.find('div', class_='success'):
+            log.info("Already logged in.")
+            return True
+
+        # Get password from keyring or prompt
+        if reenter_password is False:
+            password_from_keyring = keyring.get_password(
+                "astroquery:my.nrao.edu", username)
+        else:
+            password_from_keyring = None
+
+        if password_from_keyring is None:
+            if system_tools.in_ipynb():
+                log.warning("You may be using an ipython notebook:"
+                            " the password form will appear in your terminal.")
+            password = getpass.getpass("{0}, enter your NRAO archive password:"
+                                       "\n".format(username))
+        else:
+            password = password_from_keyring
+        # Authenticate
+        log.info("Authenticating {0} on my.nrao.edu ...".format(username))
+        # Do not cache pieces of the login process
+        data = {kw: root.find('input', {'name': kw})['value']
+                for kw in ('lt', '_eventId', 'execution')}
+        data['username'] = username
+        data['password'] = password
+        data['execution'] = 'e1s1' # not sure if needed
+        data['_eventId'] = 'submit'
+        data['submit'] = 'LOGIN'
+
+        login_response = self._request("POST", "https://my.nrao.edu/cas/login",
+                                       data=data,
+                                       cache=False)
+
+        authenticated = ('You have successfully logged in' in
+                         login_response.text)
+
+        if authenticated:
+            log.info("Authentication successful!")
+            self._username = username
```

Comment:
This seems weird, what's the difference between `self._username` and `self.USERNAME`? Also I've noticed that `USERNAME` is not defined in the class. (for comparison: ESO, https://github.com/astropy/astroquery/blob/master/astroquery/eso/core.py#L42)